crypto/evp/evp_lib.c

```diff
@@ -460,6 +460,13 @@ EVP_PKEY_CTX *EVP_MD_CTX_pkey_ctx(const EVP_MD_CTX *ctx)
     return ctx->pctx;
 }
 
+void EVP_MD_CTX_set_pkey_ctx(EVP_MD_CTX *ctx, EVP_PKEY_CTX *pctx)
```

Comment:
Good question. TBH I didn't consider the 'set to NULL' or the 'clear' scenario. But after a quick thinking, I think it would be useful to allow users to set `ctx->pctx` to NULL (for instance someone wants to reuse the MD_CTX). So the current implementation supports that already.

If `ctx->pctx` is not NULL when this function is called, then I think it should be the user's responsibility to free the old pctx attached in MD_CTX, makes sense?

Reply:
> If ctx->pctx is not NULL when this function is called, then I think it should be the user's responsibility to free the old pctx attached in MD_CTX, makes sense?

Yes, if `EVP_MD_CTX_FLAG_KEEP_PKEY_CTX` is set - but not otherwise.

Reply:
Ahh, nah, the flag should also be cleared if the `ctx->pctx` is set to NULL if someone wants to reuse the MD_CTX. So I think there should be a separated EVP_MD_CTX_clr_pkey_ctx` function to do this?

Reply:
Or we could mix the set/clr behavior into one function by doing something like:

```
if (pctx == NULL) {
    /* clear the flag... */
}
```

Reply:
Hmm..I'm not sure that is necessary. Surely all we need is something like this:

    if (!EVP_MD_CTX_test_flags(ctx, EVP_MD_CTX_FLAG_KEEP_PKEY_CTX))
        EVP_PKEY_CTX_free(ctx->pctx);
    ctx->pctx = pctx;
    if (pctx != NULL)
        EVP_MD_CTX_set_flags(ctx, EVP_MD_CTX_FLAG_KEEP_PKEY_CTX);
    else
        EVP_MD_CTX_clear_flags(out, EVP_MD_CTX_FLAG_KEEP_PKEY_CTX);

Reply:
Okay, I get the point - It's complicated to deal with such stuffs: one pointer holds two different memory management strategy objects...

Reply:
In my experience, another way is to have *two* pointers containing system-created and user-created objects respectively - if the user-created on is used, the the system one will always be NULL. Thus it would make it simpler. But this time let's just mix them together